kubernetes/resource_kubernetes_certificate_signing_request.go

```diff
@@ -0,0 +1,204 @@
+								"You can select on this field using `spec.signerName`.",
+							Optional: true,
+							ForceNew: true,
+						},
+						"usages": {
+							Type:        schema.TypeSet,
+							Description: apiDocSpec["usages"],
+							Set:         schema.HashString,
+							Elem:        &schema.Schema{Type: schema.TypeString},
+							Optional:    true,
+							ForceNew:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceKubernetesCertificateSigningRequestCreate(d *schema.ResourceData, meta interface{}) error {
+	conn, err := meta.(KubeClientsets).MainClientset()
+	if err != nil {
+		return err
+	}
+
+	metadata := expandMetadata(d.Get("metadata").([]interface{}))
+	spec, err := expandCertificateSigningRequestSpec(d.Get("spec").([]interface{}))
+	if err != nil {
+		return err
+	}
+
+	csr := v1beta1.CertificateSigningRequest{
+		ObjectMeta: metadata,
+		Spec:       *spec,
+	}
+	log.Printf("[INFO] Creating new certificate signing request: %#v", csr)
+	newCSR, createErr := conn.CertificatesV1beta1().CertificateSigningRequests().Create(&csr)
+	if createErr != nil {
+		return fmt.Errorf("Failed to create certificate signing request: %s", err)
+	}
+
+	// Get the name, since it might have been randomly generated during create.
+	csrName := newCSR.ObjectMeta.Name
+
+	if d.Get("auto_approve").(bool) {
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			pendingCSR, getErr := conn.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+			if getErr != nil {
+				return getErr
+			}
+			approval := v1beta1.CertificateSigningRequestCondition{
+				Type:    "Approved",
+				Reason:  "TerraformAutoApprove",
+				Message: "This CSR was approved by Terraform auto_approve.",
+			}
+			pendingCSR.Status.Certificate = []byte{}
+			pendingCSR.Status.Conditions = append(pendingCSR.Status.Conditions, approval)
+			_, updateErr := conn.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(pendingCSR)
+			return updateErr
+		})
+		if retryErr != nil {
+			panic(fmt.Errorf("CSR auto-approve update failed: %v", retryErr))
+		}
+		fmt.Println("CSR auto-approve update succeeded")
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Printf("[DEBUG] Waiting for certificate to be issued")
+	stateConf := &resource.StateChangeConf{
+		Target:  []string{"Issued"},
+		Pending: []string{"", "Approved"},
+		Timeout: d.Timeout(schema.TimeoutCreate),
+		Refresh: func() (interface{}, string, error) {
+			out, refreshErr := conn.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+			if refreshErr != nil {
+				log.Printf("[ERROR] Received error: %v", refreshErr)
+				return out, "Error", refreshErr
+			}
+			var csrStatus string
+			emptyStatus := v1beta1.CertificateSigningRequestStatus{}
+			emptyCSR := v1beta1.CertificateSigningRequest{}
+
+			// Ensure the CSR fields have been populated before checking them.
+			if !reflect.DeepEqual(emptyCSR, out) {
```

Comment:
I'd kind of like to invert this logic so the nesting doesn't get so deep, i.e:

```
// Ensure the CSR fields have been populated before checking them.
if reflect.DeepEqual(emptyCSR, out) {
    return out, "", nil 
}

if reflect.DeepEqual(emptyStatus, out.Status) {
    return out, "", nil
}

for _, condition := range out.Status.Conditions {
    log.Printf("[DEBUG] Found Status.Condition.Type: %v", condition.Type)
    if string(condition.Type) == "Approved" {
        // Check to see if a certificate has been issued, and update status accordingly,
        // since 'Issued' is not a state ever populated in the Status Conditions.
        if string(out.Status.Certificate) != "" {
            log.Printf("[DEBUG] Found non-empty Certificate field in Status")
            csrStatus = "Issued"
        }
    }
}
```

Reply:
Good idea, that's simpler to follow.